AsyncDisplayKit/ASViewController.mm

```diff
@@ -132,4 +142,70 @@ - (ASInterfaceState)interfaceState
   return _node.interfaceState;
 }
 
+#pragma mark - ASDisplayTraits
+
+- (ASEnvironmentDisplayTraits)displayTraitsForTraitCollection:(UITraitCollection *)traitCollection
+{
+  if (self.overrideDisplayTraitsWithTraitCollection) {
+    ASDisplayTraits *displayTraits = self.overrideDisplayTraitsWithTraitCollection(traitCollection);
+    displayTraits.isMutable = NO;
+    return [displayTraits environmentDisplayTraits];
+  }
+  
+  ASEnvironmentDisplayTraits displayTraits = ASEnvironmentDisplayTraitsFromUITraitCollection(traitCollection);
+  displayTraits.displayContext = _displayTraitsContext;
+  return displayTraits;
+}
+
+- (ASEnvironmentDisplayTraits)displayTraitsForWindowSize:(CGSize)windowSize
+{
+  if (self.overrideDisplayTraitsWithWindowSize) {
+    ASDisplayTraits *displayTraits = self.overrideDisplayTraitsWithWindowSize(windowSize);
+    displayTraits.isMutable = NO;
+    return [displayTraits environmentDisplayTraits];
+  }
+  return self.node.environmentState.displayTraits;
+}
+
+- (void)progagateNewDisplayTraits:(ASEnvironmentDisplayTraits)displayTraits
+{
+  ASEnvironmentState environmentState = self.node.environmentState;
+  ASEnvironmentDisplayTraits oldDisplayTraits = environmentState.displayTraits;
+  
+  if (ASEnvironmentDisplayTraitsIsEqualToASEnvironmentDisplayTraits(displayTraits, oldDisplayTraits) == NO) {
+    environmentState.displayTraits = displayTraits;
+    [self.node setEnvironmentState:environmentState];
+    [self.node setNeedsLayout];
+    
+    NSArray<id<ASEnvironment>> *children = [self.node children];
+    for (id<ASEnvironment> child in children) {
+      ASEnvironmentStatePropagateDown(child, environmentState.displayTraits);
+    }
+  }
+}
+
+- (void)traitCollectionDidChange:(UITraitCollection *)previousTraitCollection
+{
+  [super traitCollectionDidChange:previousTraitCollection];
+  
+  ASEnvironmentDisplayTraits displayTraits = [self displayTraitsForTraitCollection:self.traitCollection];
+  [self progagateNewDisplayTraits:displayTraits];
+}
+
+- (void)willTransitionToTraitCollection:(UITraitCollection *)newCollection withTransitionCoordinator:(id<UIViewControllerTransitionCoordinator>)coordinator
+{
+  [super willTransitionToTraitCollection:newCollection withTransitionCoordinator:coordinator];
+  
+  ASEnvironmentDisplayTraits displayTraits = [self displayTraitsForTraitCollection:self.traitCollection];
+  [self progagateNewDisplayTraits:displayTraits];
+}
+
+- (void)viewWillTransitionToSize:(CGSize)size withTransitionCoordinator:(id<UIViewControllerTransitionCoordinator>)coordinator
```

Comment:
Why do we need to implement all three of these notifications? `willTransitionToTraitCollection:` should suffice, as it is called in `transitionToSize:` call and is invoked before `traitCollectionDidChange:` is called. I'd love to learn more about the reasons behind this implementation.